Advanced logging config options in google_compute_subnetwork

dustin-decker opened the pull request:
Added custom metadata fields and filter expressions to subnetwork flow log configuration. Fixes https://github.com/hashicorp/terraform-provider-google/issues/6500

Signed-off-by: Dustin Decker <dustin.decker@getcruise.com>

<!-- AUTOCHANGELOG for Downstream PRs.

Please select one of the following "release-note:" headings:
    - release-note:note
    
Unless you choose release-note:none, please add a release note.

See .ci/RELEASE_NOTES_GUIDE.md for writing good release notes.

You can add more release note blocks if you want more than one CHANGELOG
entry for this PR.
-->

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
compute: Added custom metadata fields and filter expressions to `google_compute_subnetwork` flow log configuration
```

modular-magician commented:
Hello!  I am a robot who works on Magic Modules PRs.

I have detected that you are a community contributor, so your PR will be assigned to someone with a commit-bit on this repo for initial review.

Thanks for your contribution!  A human will be with you soon.

@danawillow, please review this PR or find an appropriate assignee.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3603-old..auto-pr-3603) ( 2 files changed, 25 insertions(+), 3 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3603-old..auto-pr-3603) ( 2 files changed, 25 insertions(+), 3 deletions(-))
Ansible: [Diff](https://github.com/modular-magician/google.cloud/compare/auto-pr-3603-old..auto-pr-3603) ( 1 file changed, 1 insertion(+), 1 deletion(-))
Inspec: [Diff](https://github.com/modular-magician/inspec-gcp/compare/auto-pr-3603-old..auto-pr-3603) ( 2 files changed, 11 insertions(+))

modular-magician commented:
Oops! It looks like you're using an unknown release-note type in your changelog entries:
* 

Please only use the types listed in https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md.

danawillow commented:
This fixes https://github.com/terraform-providers/terraform-provider-google/issues/6500,  right? Can you put that in the PR description?

I'll take a look once the CLA is resolved, but my comments are almost certainly going to be "can you add a test for this?"

dustin-decker commented:
Yes, it should resolve https://github.com/terraform-providers/terraform-provider-google/issues/6500

Thanks for mentioning, I didn't see the issue before. I'll look at adding a test.

dustin-decker commented:
@googlebot I signed it!

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3603-old..auto-pr-3603) ( 3 files changed, 65 insertions(+), 5 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3603-old..auto-pr-3603) ( 3 files changed, 65 insertions(+), 5 deletions(-))
Ansible: [Diff](https://github.com/modular-magician/google.cloud/compare/auto-pr-3603-old..auto-pr-3603) ( 1 file changed, 1 insertion(+), 1 deletion(-))
Inspec: [Diff](https://github.com/modular-magician/inspec-gcp/compare/auto-pr-3603-old..auto-pr-3603) ( 2 files changed, 11 insertions(+))

dustin-decker commented:
Added terraform test

danawillow commented on third_party/terraform/tests/resource_compute_subnetwork_test.go:
```diff
@@ -577,6 +589,32 @@ resource "google_compute_subnetwork" "network-with-flow-logs" {
+  log_config {
+    aggregation_interval = "INTERVAL_30_SEC"
+    flow_sampling        = 0.8
+	metadata             = "CUSTOM_METADATA"
```
watch out for tabs vs spaces

dustin-decker replied:
fixed

danawillow commented on products/compute/api.yaml:
```diff
@@ -12966,14 +12966,24 @@ objects:
+          - !ruby/object:Api::Type::Array
+            name: 'metadataFields'
+            description: |
+              Can only be specified if VPC flow logs for this subnetwork is enabled and "metadata" was set to CUSTOM_METADATA.
```
```suggestion
              Can only be specified if VPC flow logs for this subnetwork is enabled and "metadata" is set to CUSTOM_METADATA.
```

dustin-decker replied:
Changed

danawillow commented on products/compute/api.yaml:
```diff
@@ -12966,14 +12966,24 @@ objects:
               - :INCLUDE_ALL_METADATA
+              - :CUSTOM_METADATA
             default_value: :INCLUDE_ALL_METADATA
+          - !ruby/object:Api::Type::Array
```
Does order matter here? If not, I'd recommend marking this as `is_set: true` in terraform.yaml

dustin-decker replied:
Set this value to make it treated as a set

modular-magician commented:
Hello!  I am a robot who works on Magic Modules PRs.

I have detected that you are a community contributor, so your PR will be assigned to someone with a commit-bit on this repo for initial review.

Thanks for your contribution!  A human will be with you soon.

@c2thorn, please review this PR or find an appropriate assignee.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3603-old..auto-pr-3603) ( 3 files changed, 77 insertions(+), 5 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3603-old..auto-pr-3603) ( 3 files changed, 77 insertions(+), 5 deletions(-))
Ansible: [Diff](https://github.com/modular-magician/google.cloud/compare/auto-pr-3603-old..auto-pr-3603) ( 1 file changed, 1 insertion(+), 1 deletion(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3603-old..auto-pr-3603) ( 1 file changed, 10 insertions(+))
Inspec: [Diff](https://github.com/modular-magician/inspec-gcp/compare/auto-pr-3603-old..auto-pr-3603) ( 2 files changed, 11 insertions(+))

modular-magician commented:
I have triggered VCR tests based on this PR's diffs. See the results here: "https://ci-oss.hashicorp.engineering/viewQueued.html?itemId=119008"

dustin-decker commented:
I've incorporated your suggestions and added the custom expanders and flatteners. Thanks for the pointers, there is a lot to get familiar with in this repo. Tests are passing on my machine.

danawillow commented on templates/terraform/custom_expand/subnetwork_log_config.go.erb:
```diff
@@ -35,6 +35,14 @@ func expand<%= prefix -%><%= titlelize_property(property) -%>(v interface{}, d T
+	transformed["metadataFields"] = original["metadata_fields"].(*schema.Set).List()
+
+	// allow a null default (must eval to true otherwise)
+	if val := reflect.ValueOf(original["filter_expr"]); val.IsValid() && !isEmptyValue(val) {
```
I'm  not 100% sure, but I think this is going to mean that `filter_expr` will not be able to be updated to an empty value after its been set (it depends on how the API handles unset fields in the PATCH request). Can you check?

dustin-decker replied:
I've added a test that I hope covers unsetting this and the metadata fields. The API did not accept an empty value, but would accept a unset value.

dustin-decker replied:
Oops.. test is failing. I'll take a closer look.

dustin-decker replied:
I'm having some trouble with this. 

If it's unset, I think it is not updating.

If it sends a empty string it gets this error:

```
Error: Error creating Subnetwork: googleapi: Error 400: Invalid value for field 'resource.logConfig.filterExpr': ''. ERROR: VPC Flow Log filter expression:1:0: mismatched input '<EOF>' expecting {'[', '{', '(', '.', '-', '!', 'true', 'false', 'null', NUM_FLOAT, NUM_INT, NUM_UINT, STRING, BYTES, IDENTIFIER}
    | 
    | ^
ERROR: VPC Flow Log filter expression:1:0: Filter expression does not evaluate to a boolean
    | 
    | ^, invalid

    on /tmp/tf-test941063018/main.tf line 7:
    (source code not available)
```

Is the solution to send the filter expression string `true` when it's unset? That feels wrong.

danawillow replied:
That does feel wrong. If you have the exact request/response from GCP (which you can get by running the test with env var TF_LOG=DEBUG) I can help look into it.

modular-magician commented:
Hello!  I am a robot who works on Magic Modules PRs.

I have detected that you are a community contributor, so your PR will be assigned to someone with a commit-bit on this repo for initial review.

Thanks for your contribution!  A human will be with you soon.

@ndmckinley, please review this PR or find an appropriate assignee.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3603-old..auto-pr-3603) ( 3 files changed, 110 insertions(+), 5 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3603-old..auto-pr-3603) ( 3 files changed, 110 insertions(+), 5 deletions(-))
Ansible: [Diff](https://github.com/modular-magician/google.cloud/compare/auto-pr-3603-old..auto-pr-3603) ( 1 file changed, 1 insertion(+), 1 deletion(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3603-old..auto-pr-3603) ( 1 file changed, 10 insertions(+))
Inspec: [Diff](https://github.com/modular-magician/inspec-gcp/compare/auto-pr-3603-old..auto-pr-3603) ( 2 files changed, 11 insertions(+))

modular-magician commented:
I have triggered VCR tests based on this PR's diffs. See the results here: "https://ci-oss.hashicorp.engineering/viewQueued.html?itemId=120784"

danawillow commented:
I played around with this a bit today, and I think it's a bug in the API. I filed a bug against the team (b/162771344 for googlers), and will keep you posted on progress! If it ends up taking too long to resolve, we could also just mark the field as not updatable in order to get the functionality in.

dustin-decker commented:
Thanks Dana! Sorry I let this sit so long. I suspected the same and put off trying to thoroughly prove it. I'll be happy to pick it back up when there's a path forward.

danawillow commented:
All right, I heard back from the API team and their recommendation was just to default to "true", since that's equivalent to leaving it empty. I'm now trying to find out whether it's possible to set `metadataFields` to empty as well, since that seemed to fail when I ran the tests locally.

danawillow commented:
Ok, they said that for metadataFields clearing it is equivalent to setting metadata to EXCLUDE_ALL_METADATA. I haven't actually tried it in the API yet to see if that clears the field or not. Let me know if you want me to take over any of this! I'd love to get it in soon.

dustin-decker commented:
I'm busy for the next couple of weeks so if you'd like to take it the rest of the way I'd be grateful. We're looking forward to this feature too.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3603-old..auto-pr-3603) ( 3 files changed, 124 insertions(+), 9 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3603-old..auto-pr-3603) ( 3 files changed, 124 insertions(+), 9 deletions(-))
Ansible: [Diff](https://github.com/modular-magician/google.cloud/compare/auto-pr-3603-old..auto-pr-3603) ( 1 file changed, 1 insertion(+), 1 deletion(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3603-old..auto-pr-3603) ( 1 file changed, 6 insertions(+))
Inspec: [Diff](https://github.com/modular-magician/inspec-gcp/compare/auto-pr-3603-old..auto-pr-3603) ( 2 files changed, 11 insertions(+))
